src/sensors/IMD.py

```diff
@@ -0,0 +1,142 @@
+from test_lib.condition.condition import Condition
+from test_lib.input.timed import After
+from enum import Enum
+
+
+class Case(Enum):
+    ShortCircuit = 0
+    Normal = 10
+    UnderVoltage = 20
+    EquipmentFault = 40
+    GroundFault = 50
+
+
+class IMD:
+    def __init__(
+        self, shm: SharedMemory, ok_hs: Pinout, m_hs: Pinout, power_on: Pinout
+    ):
+        self.power_on = DigitalOutService(shm, power_on)
+        self.ok_hs = DigitalInService(shm, ok_hs)
+        self.m_hs = InputCapture(shm, m_hs)
+
+    def _IMDInput(self, frequency, duty_cycle, good_state: bool) -> Input:
+        if self.power_on.get_pin_state() == memory.DigitalOut.State.Low:
+            raise InputFailedException
+        elif good_state is True:
+            return Multiple(
+                self.ok_hs.generate_state(memory.DigitalIn.State.High),
+                self.m_hs.generate_duty(duty_cycle),
+                self.m_hs.generate_frequency(frequency),
+            )
+        else:
+            return Multiple(
+                self.ok_hs.generate_state(memory.DigitalIn.State.Low),
+                self.m_hs.generate_duty(duty_cycle),
+                self.m_hs.generate_frequency(frequency),
+            )
+
+    def check_is_on(self) -> Condition:
+        return self.power_on.wait_for_high()
+
+    def generate_fast_start(self, duty_cycle) -> Input:
+        if 5 <= duty_cycle <= 10:
+            return self._IMDInput(30, duty_cycle, True)
+        if 90 <= duty_cycle <= 95:
+            return self._IMDInput(30, duty_cycle, False)
+        else:
+            raise ValueError(
+                "Duty cycle for fast start can be only between 5%% and 10%%, or between 90%% and 95%%"
+            )
+
+    def generate_short_circuit(self) -> Input:
+        return self._IMDInput(0, 0, 0)
+
+    def generate_normal(self, duty_cycle) -> Input:
+        if 5 <= duty_cycle <= 95:
+            raise ValueError("Duty cycle for normal state must be between 5%% and 95%%")
+        return self._IMDInput(10, duty_cycle, True)
+
+    def generate_undervoltage(self, duty_cycle) -> Input:
+        if not 5 <= duty_cycle <= 95:
+            raise ValueError(
+                "Duty cycle for undervoltage state must be between 5%% and 95%%"
+            )
+        return self._IMDInput(20, duty_cycle, False)
+
+    def generate_equipment_fault(self, duty_cycle) -> Input:
+        if not 47.5 <= duty_cycle <= 52.5:
+            raise ValueError(
+                "Duty cycle for equipment fault must be between 47.5%% and 52.5%%"
+            )
+        return self._IMDInput(40, duty_cycle, False)
+
+    def generate_ground_fault(self, duty_cycle) -> Input:
+        if not 47.5 <= duty_cycle <= 52.5:
+            raise ValueError(
+                "Duty cycle for ground fault must be between 47.5%% and 52.5%%"
+            )
+        return self._IMDInput(50, duty_cycle, False)
+
+    def generate_power_up(self, case: Case, initial_duty_cycle, duty_cycle) -> Input:
```

Comment:
TL;DR: Have the user provide the resistance (in ohms, kohms, or whatever magnitude) and do the conversion to duty cycle inside the function

instead of having the user calculate the duty cycle in a previous step for their particular case, I think it is more intuitive to ask for the resistance, that way you hide from the user the inner workings of the PWM and you just have to make the conversion inside, something that would be needed anyways.

Reply:
I can do it with duty cycle after the two initial seconds, but I can't do it with the initial duty cycle because the IMD cheatseet doesn't specify the relation between this duty cycle and the resistance. The formula used to do the conversion in resistance_to_duty_cycle() function only applied to "stable" cases, that means after the two initial seconds

Reply:
I don't have the datasheet at hand, but IIRC the module ships with a factory configured threshold for what is good or bad, which I think was 500 ohms for our model. I guess a good duty cycle would be when the resistance measured is over 500 ohms and bad otherwise